xwayland/xwayland.c

```diff
@@ -0,0 +1,227 @@
+#define _XOPEN_SOURCE 700
+#define _GNU_SOURCE
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/socket.h>
+#include <time.h>
+#include <errno.h>
+#include <wayland-server.h>
+#include "wlr/util/log.h"
+#include "wlr/xwayland.h"
+#include "sockets.h"
+#include "xwm.h"
+
+static void safe_close(int fd) {
+	if (fd >= 0) {
+		close(fd);
+	}
+}
+
+static int unset_cloexec(int fd) {
+	if (fcntl(fd, F_SETFD, 0) != 0) {
+		wlr_log_errno(L_ERROR, "fcntl() failed on fd %d", fd);
+		return -1;
+	}
+	return 0;
+}
+
+static void exec_xwayland(struct wlr_xwayland *wlr_xwayland) {
+	if (unset_cloexec(wlr_xwayland->x_fd[0]) ||
+			unset_cloexec(wlr_xwayland->x_fd[1]) ||
+			unset_cloexec(wlr_xwayland->wm_fd[1]) ||
+			unset_cloexec(wlr_xwayland->wl_fd[1])) {
+		exit(EXIT_FAILURE);
+	}
+
+	/* Make Xwayland signal us when it's ready */
+	signal(SIGUSR1, SIG_IGN);
+
+	char *argv[11] = { 0 };
```

Comment:
Yeah I actually agonized a bit over this one. I thought of that but didn't like doing the argument that need formating out of order because argv[4] (for example) could move easily.
Hmm, now I've slept on it maybe place-hold with NULL and fill in there to automatically adjust if someone adds static strings. Will come up with something, and remove asprintf (that's the gnu extension here)